plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastoreModule.java

```diff
@@ -71,4 +77,17 @@ public HiveMetastore createCachingHiveMetastore(
 
         return cachingHiveMetastore(decoratedDelegate, executor, config);
     }
+
+    @Provides
+    @Singleton
+    public Optional<CachingHiveMetastore> createOptionalCachingHiveMetastore(
+            NodeManager nodeManager,
+            CachingHiveMetastoreConfig config,
+            HiveMetastore hiveMetastore)
+    {
+        if (!nodeManager.getCurrentNode().isCoordinator() || isCacheDisabled(config)) {
+            return Optional.empty();
+        }
+        return Optional.of((CachingHiveMetastore) hiveMetastore);
```

Comment:
That looks like indirect way to construct `Optional<CachingHiveMetastore>`. Can we make `createCachingHiveMetastore` depend on `createOptionalCachingHiveMetastore`.
Perhaps`io.trino.plugin.hive.metastore.cache.CachingHiveMetastore#cachingHiveMetastore` should return `Optional`.
This way type usage is more explicit and safer

Reply:
I was not able to find a way to not affect current expected behaviour.
Everything just binds HiveMetastore not knowing which implementation will be injected. 

Current implementation is returning one of hive meta store implementations depending on configuration or decorating it (without usage of Guice - indirectly) with CachingHiveMetastore.
Here we return non cached version https://github.com/trinodb/trino/blob/bcaf0547f3a57fe899d4669aac3cc78d6c4af53b/plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastoreModule.java#L65
and here cached
https://github.com/trinodb/trino/blob/bcaf0547f3a57fe899d4669aac3cc78d6c4af53b/plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastoreModule.java#L72
But it may also happen that even this last location which return non cached version here:
https://github.com/trinodb/trino/blob/bcaf0547f3a57fe899d4669aac3cc78d6c4af53b/plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java#L136

That's why i preserved this when creating Optional in same way just centralised decision point.

Reply:
> I was not able to find a way to not affect current expected behaviour.
Everything just binds HiveMetastore not knowing which implementation will be injected.

You can still decorate.
```
createOptionalCachingHiveMetastore
```
returns either decorated metastore or `empty`.

Then
```
createCachingHiveMetastore
```
would either return `CachingMetastore` or original one.